### DISCUSSION TITLE
How can I add ResourceDictionary by code?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#16569] [Q&A] [answered]
I want to add different ResourceDictionary (for templatedcontrol) according to different screen orientation.

So, I need to use code to implement this.

ResourceDictionary Sample:
```
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
	<ResourceDictionary.MergedDictionaries>
		<ResourceInclude Source="avares://Sample/One.axaml"/>
		<ResourceInclude Source="avares://Sample/Two.axaml" />
	</ResourceDictionary.MergedDictionaries>
</ResourceDictionary>
```

And here is my App.axaml:
```
<Application xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             x:Class="Sample.App"
             RequestedThemeVariant="Default">
	<!-- "Default" ThemeVariant follows system theme variant. "Dark" or "Light" are other available options. -->

	<Application.Styles>
		<FluentTheme />
		
	</Application.Styles>
	<Application.Resources>
		<ResourceDictionary>
			<ResourceDictionary.MergedDictionaries>
				
			</ResourceDictionary.MergedDictionaries>
		</ResourceDictionary>
	</Application.Resources>
</Application>
```

And here is my code:
```
var _stream = AvaloniaXamlLoader.Load(new Uri(uri));
if(_stream is ResourceDictionary _resourceDictionary)
     {
         if (Application.Current is { })
         {
             foreach (var j in _resourceDictionary)
             {
               Application.Current.Resources.Add(j);               
             }
         }     
     }
```

It runs without any exception while the templatedcontrol is blank in visual tree. It seems the ResourceDictionary did not add correctly at all.

Is there something wrong with my code? How can I correctly add a ResourceDictionary via code?

**Comment by maxkatz6** [comment, marked as answer]
You do it the same way how it is in XAML.
1. Create ResourceInclude and set Source property.
2. Add previously created ResourceInclude to the Application.Current.Resources.MergedDictionaries

**Reply by maxkatz6** [reply]
No need for AvaloniaXamlLoader.Load, although you can use that too alternatively

**Reply by kyuranger** [reply]
> No need for AvaloniaXamlLoader.Load, although you can use that too alternatively

Initially, I tried add it like this:
```
ResourceInclude _resourceInclude = new ResourceInclude(new Uri("avares://Sample/One.axaml"));
Application.Current.Resources.MergedDictionaries.Add(_resourceInclude);
```

Whereas, VS reports an error: 
`System.InvalidOperationException: 'ResourceInclude.Source must be set.'`

I don't know why I have declared the URI before but it said I never set. So, I changed my code like this:
```
ResourceInclude _resourceInclude = new ResourceInclude(new Uri("avares://Sample/One.axaml")){ Source=new Uri("avares://Sample/One.axaml") };;
Application.Current.Resources.MergedDictionaries.Add(_resourceInclude);
```

It runs without any error now while the visual tree of the templatedcontrol is still blank.

I consider there must something wrong that I declared the URI twice.

**Reply by maxkatz6** [reply]
> It runs without any error now while the visual tree of the templatedcontrol is still blank.

Your second code is correct-ish. The important part is `Source` property, it needs to be set. Uri in the constructor is a base Uri, used when Source is relative. This weird API is a left-over after old XAML integration. 
BaseUri can be null, if you prefer.

> I consider there must something wrong that I declared the URI twice.

There should be something wrong elsewhere. Even with duplicated Uri.

**Reply by kyuranger** [reply]
> > It runs without any error now while the visual tree of the templatedcontrol is still blank.
> 
> Your second code is correct-ish. The important part is `Source` property, it needs to be set. Uri in the constructor is a base Uri, used when Source is relative. This weird API is a left-over after old XAML integration. BaseUri can be null, if you prefer.
> 
> > I consider there must something wrong that I declared the URI twice.
> 
> There should be something wrong elsewhere. Even with duplicated Uri.

When I set the BaseUri to null, VS reports an error that it cannot be null.

I am a little suspicious whether the method of using `Application.Current.Resources.MergedDictionaries` is wrong. Because when I add a breakpoint in the loaded event of the window and monitor the content added in `Application.Current.Resources.MergedDictionaries`, I can't find the Resource I added to other templatedcontrols through APP.AXAML (but other templatedcontrols are running normally).

**Reply by kyuranger** [reply]
> > It runs without any error now while the visual tree of the templatedcontrol is still blank.
> 
> Your second code is correct-ish. The important part is `Source` property, it needs to be set. Uri in the constructor is a base Uri, used when Source is relative. This weird API is a left-over after old XAML integration. BaseUri can be null, if you prefer.
> 
> > I consider there must something wrong that I declared the URI twice.
> 
> There should be something wrong elsewhere. Even with duplicated Uri.

I found some problems with my other codes. The method you provided is feasible. The code should be:
```
ResourceInclude _resourceInclude = new ResourceInclude(new Uri("avares://Sample/One.axaml")){ Source=new Uri("avares://Sample/One.axaml") };;
Application.Current.Resources.MergedDictionaries.Add(_resourceInclude);
```

**Reply by maxkatz6** [reply]
> When I set the BaseUri to null, VS reports an error that it cannot be null.

My bad.
But since it's a "base uri", you can put something as "avares://Sample/" keeping it simple. Source then can be a "/One.axaml". At least that's how it works with XAML, where users mostly use relative paths.

**Reply by kyuranger** [reply]
Thanks a lot.
